server/src/main/java/org/opensearch/index/engine/Engine.java

```diff
@@ -1309,6 +1326,35 @@ public void failEngine(String reason, @Nullable Exception failure) {
         }
     }
 
+    /**
+     * Cleanup all unreferenced files generated during failed segment merge. This resets shard state to last Lucene
+     * commit.
+     */
+    private void cleanUpUnreferencedFiles() {
+        try (
+            IndexWriter writer = new IndexWriter(
+                store.directory(),
+                new IndexWriterConfig(Lucene.STANDARD_ANALYZER).setSoftDeletesField(Lucene.SOFT_DELETES_FIELD)
+                    .setCommitOnClose(false)
+                    .setMergePolicy(NoMergePolicy.INSTANCE)
+                    .setOpenMode(IndexWriterConfig.OpenMode.APPEND)
+            )
+        ) {
+            // do nothing and close this will kick off IndexFileDeleter which will remove all unreferenced files
+        } catch (Exception ex) {
+            logger.error("Error while deleting unreferenced file ", ex);
+        }
+    }
+
+    private boolean isMergeFailureDueToDiskFull(Exception failure, String reason) {
+        return (reason.equals(FORCE_MERGE) || reason.equals(MERGE_FAILED))
+            && failure != null
+            && failure.getCause() != null
+            && failure.getCause().getCause() != null
+            && failure.getCause().getCause().getMessage() != null
+            && failure.getCause().getCause().getMessage().contains("No space left on device");
+    }
```

Comment:
This check on checking message looks weak? What happens if we do this on IOException

Reply:
Consume ExceptionsHelper utilities for stuff like these

Reply:
> What happens if we do this on IOException

Clenaup needs to be only on disk full. For other IOException like ```READ_ONLY_FILE_SYSTEM``` disks we should not perform cleanup. Also will refactor this using ExceptionsHelper utilities.

Reply:
Addressed the comment

Reply:
Had internal discussions on this. It seems doing a best effort cleanup on IOException seems much more sense than parsing exception messages for disk full. Adressed the comments.